### PR TITLE
IdCompressor Telemetry 

### PR DESCRIPTION
**Pull request by justus-camp**
# [Task 560](https://dev.azure.com/fluidframework/internal/_workitems/edit/560)

## Description

- Adding telemetry for the IdCompressor to justify the use of an automatically resizing cluster scheme.
- The intention is for these telemetry events to give us information on how many eager final IDs are being allocated and how often new clusters are being created.
- I'm keeping the events in finalizeCreationRange to limit the overhead of emitting these events.

I'm not convinced I've covered what I'm intending to with these events. Because both local and remote sessions come through this method, we want to make sure that we don't mix local vs. remote data. 

## Does this introduce a breaking change?

This is technically a breaking change, but the IdCompressor is only being used internally in SharedTree. Adding a logger to the IdCompressor shouldn't have an impact on anyone.

### CODE REVIEW

File: experimental/dds/tree/src/id-compressor/IdCompressor.ts [hunk 1]
```diff
@@ -384,13 +387,15 @@ export class IdCompressor {
 	public constructor(
 		public readonly localSessionId: SessionId,
 		public readonly reservedIdCount: number,
-		attributionId?: AttributionId
+		attributionId?: AttributionId,
+		logger?: ITelemetryLogger
```
Comment (taylorsw04) [inline experimental/dds/tree/src/id-compressor/IdCompressor.ts:391]: should be able to make this line something like `private readonly logger?: ItelemetryLogger` and avoid the field declaration and assignment

**Review by taylorsw04 — CHANGES_REQUESTED** [review]

File: experimental/dds/tree/src/id-compressor/IdCompressor.ts [hunk 2]
```diff
@@ -384,13 +385,15 @@ export class IdCompressor {
...
-		attributionId?: AttributionId
+		attributionId?: AttributionId,
+		private readonly logger?: ITelemetryLogger
 	) {
 		assert(reservedIdCount >= 0, 'reservedIdCount must be non-negative');
 		if (attributionId !== undefined) {
 			assertIsUuidString(attributionId);
 		}
 		this.localSession = this.createSession(localSessionId, attributionId);
+		this.logger = logger;
```
Comment (taylorsw04) [inline experimental/dds/tree/src/id-compressor/IdCompressor.ts:396]: You don't need this assignment anymore :)

Comment (justus-camp) [reply experimental/dds/tree/src/id-compressor/IdCompressor.ts]: 😮

**Comment by justus-camp** [comment]
Took a stab at what Noah and I talked about yesterday. Left the events in with less info and then created an aggregate event that outputs what we're interested in. I left some data in the expansion case and also the new cluster case so that we can look at those in more detail. I added a field for override count but I'm not sure if it's correct? Worth taking a look at each metric to ensure correctness. Once we're happy with these I'll fix the tests and move this out of draft state.

**Comment by justus-camp** [comment]
I think all the cases you listed are covered now besides the attribution info creation. Is there value in including the size of attribution info? As far as I understand it won't it be a UUID string (which should all have the same length?). I guess the actual value would be that there is attribution info present?

**Comment by taylorsw04** [comment]
@justus-camp whoops I forgot that attribution info now has a bounded size. So I actually don't see much value there at all.

**Comment by taylorsw04** [comment]
The rest looks good to me :)

**Review by taylorsw04 — APPROVED** [review]

**Pull request merged**